libraries/AP_InertialSensor/AP_InertialSensor_tempcal.cpp

```diff
@@ -0,0 +1,491 @@
+ */
+void AP_InertialSensor::TCal::correct_sensor(float temperature, float cal_temp, const AP_Vector3f coeff[3], Vector3f &v) const
+{
+    if (enable != Enable::Enabled) {
+        return;
+    }
+    temperature = constrain_float(temperature, temp_min, temp_max);
+    cal_temp = constrain_float(cal_temp, temp_min, temp_max);
+
+    const float tmid = (temp_max + temp_min)*0.5;
+    if (tmid <= 0) {
+        return;
+    }
+
+    // get the polynomial correction for the difference between the
+    // current temperature and the mid temperature
+    v -= polynomial_eval(temperature - tmid, coeff);
+
+    // we need to add the correction for the temperature
+    // difference between the tmid, which is the reference used for
+    // the calibration process, and the cal_temp, which is the
+    // temperature that the offsets and scale factors was setup for
+    v += polynomial_eval(cal_temp - tmid, coeff);
+}
+
+void AP_InertialSensor::TCal::correct_accel(float temperature, float cal_temp, Vector3f &accel) const
+{
+    correct_sensor(temperature, cal_temp, accel_coeff, accel);
+}
+
+void AP_InertialSensor::TCal::correct_gyro(float temperature, float cal_temp, Vector3f &gyro) const
+{
+    correct_sensor(temperature, cal_temp, gyro_coeff, gyro);
+}
+
+void AP_InertialSensor::TCal::sitl_apply_accel(float temperature, Vector3f &accel) const
+{
+    Vector3f v;
+    correct_sensor(temperature, 0.5*(temp_max+temp_min), accel_coeff, v);
+    accel -= v;
+}
+
+void AP_InertialSensor::TCal::sitl_apply_gyro(float temperature, Vector3f &gyro) const
+{
+    Vector3f v;
+    correct_sensor(temperature, 0.5*(temp_max+temp_min), gyro_coeff, v);
+    gyro -= v;
+}
+
+AP_InertialSensor::TCal::Learn::Learn(TCal &_tcal, float _start_temp) :
+    start_temp(_start_temp),
+    tcal(_tcal)
+{
+    reset(_start_temp);
+}
+
+/*
+  update polyfit with new sample
+ */
+void AP_InertialSensor::TCal::Learn::add_sample(const Vector3f &sample, float temperature, struct LearnState &st)
+{
+    temperature = st.temp_filter.apply(temperature);
+
+    st.sum += sample;
+    st.sum_count++;
+
+    if (st.sum_count < 100 ||
+        temperature - st.last_temp < 0.5) {
+        // wait for more data
+        return;
+    }
+
+    st.sum /= st.sum_count;
+
+    const uint8_t si = &st - &state[0];
+
+    const float T = (temperature + st.last_temp) * 0.5;
+
+    if (si == 0) {
+        // we use the first accel sample as the zero baseline
+        if (accel_start.is_zero()) {
+            accel_start = st.sum;
+            start_temp = T;
+        }
+        st.sum -= accel_start;
+    }
+    
+    const float tmid = 0.5 * (tcal.temp_max + start_temp);
+    const float tdiff = T - tmid;
+
+    AP::logger().Write("TCLR", "TimeUS,I,Si,Temp,TDiff,X,Y,Z",
+                       "s#------",
+                       "F0000000",
+                       "QBBfffff",
+                       AP_HAL::micros64(),
+                       instance(),
+                       si,
```

Comment:
Its not immediately clear as to what si represent, either a comment or better name would be helpful. It seems it represents Accel(0) or Gyro(1) of the IMU, right? Also can we add number of samples averaged here as well?

Reply:
I've changed it to SType, and yes, 0 is accel, 1 for gyro
I've added NSamp as well